chapter2/testing.md

```diff
@@ -0,0 +1,172 @@
+# Testing
+
+We strongly encourage contributors to test the elements they create.
+We usually advise on creating two types of tests: Unit Tests and Integration Tests.
+To facilitate writing tests and make the developer experience as smooth as possible, we expose a couple of testing utilities inside `Membrane.Testing` package.
+
+We will discuss and walk you through writing tests on a real-world example for the [AAC decoder](https://github.com/membraneframework/membrane-element-aac) element.
+
+## Setting up
+
+In order to simplify things and make the test easier to maintain and modify, we propose the following solution:
+
+- Create a sample file for the element under test. As an example, for the AAC decoder we'll use a very simple AAC audio file
+- Create a known output as a reference file. For example, we save raw audio frames as the reference output for our AAC Decoder. **Note**: This may not always be the best solution since most encoders produce a non-deterministic output for the same input
+- Use the input file and reference file as sources of truth for your tests
+
+## Running tests
+
+In order to run all tests, execute:
+
+```sh
+$ mix test
+```
+
+## Unit Tests
+
+We will create unit tests for our Native decoder.
+
+We start with a standard test module definition
+
+```elixir
+defmodule Decoder.NativeTest do
+  use ExUnit.Case, async: true
+  alias Membrane.Element.AAC.Decoder.Native
+```
+
+Our test scenario will check if the native decoder is able to decode a single frame of data:
+
+```elixir
+  test "Decode 1 AAC frame" do
+```
+
+As described above, we will use input/reference files as our data sources. Those files are placed inside `test/fixtures` directory
+
+```elixir
+    in_path = "fixtures/input-sample.aac" |> Path.expand(__DIR__)
+    reference_path = "fixtures/reference-sample.raw" |> Path.expand(__DIR__)
+    ...
+    assert {:ok, file} = File.read(in_path)
+```
+
+The first step is to ensure we can create a `Native` decoder:
+
+```elixir
+    assert {:ok, decoder_ref} = Native.create()
+```
+
+Next, we will retrieve the first frame from the input file, pass it to the Native decoder and assert that it was decoded correctly. Subsequent calls to `decode_frame` should return `:not_enough_bits`, as all data is already parsed.
+
+```elixir
+    assert <<frame::bytes-size(256), _::binary>> = file
+    assert :ok = Native.fill(frame, decoder_ref)
+    assert {:ok, decoded_frame} = Native.decode_frame(frame, decoder_ref)
+    assert {:error, :not_enough_bits} = Native.decode_frame(frame, decoder_ref)
+```
+
+Finally, we'll compare the decoded_frame with a reference frame from the saved raw file.
```

Comment:
If you meant variable `decoded_frame` then add backticks, if not - replace underscore with space